index/index.go

```diff
@@ -0,0 +1,94 @@
+package index
+
+import (
+	"encoding/json"
+	"io"
+	"sync"
+)
+
+type Index interface {
+	Match(terms []string) []string
+	Add(doc string, terms []string)
+	WriteTo(w io.Writer) error
+}
+
+type MapIndex struct {
+	mu    sync.RWMutex
+	Docs  []string
+	Terms map[string][]int
+}
+
+func ReadMapIndex(r io.Reader) (*MapIndex, error) {
+	dec := json.NewDecoder(r)
+	ind := &MapIndex{}
+	if err := dec.Decode(ind); err != nil {
+		return nil, err
+	}
+	return ind, nil
+}
+
+func (i *MapIndex) Match(terms []string) []string {
+	i.mu.RLock()
+	defer i.mu.RUnlock()
+
+	var docsLists [][]int
+	for _, t := range terms {
+		docs := i.Terms[t]
+		if len(docs) > 0 {
+			docsLists = append(docsLists, docs)
+		}
+	}
+	if len(docsLists) == 0 {
+		return nil
+	}
+
+	positions := make([]int, len(docsLists))
+	var results []string
+	for {
+		min := len(i.Docs)
+		max := -1
+		for i, pos := range positions {
```

Comment:
Have stared at this for quite a while now and still can't quite work out what this does (or is supposed to do).

For each term, we take the min and max matching doc, ... and then?

Would it be possible to add a comment to the Match interface method explaining what it returns and/or give a few of these variables more descriptive names?